internal/worker/dbreplaccessor/tracker.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package dbreplaccessor
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v4"
+	"gopkg.in/tomb.v2"
+
+	corecontext "github.com/juju/juju/core/context"
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/database/pragma"
+)
+
+// TrackedDB defines the union of a TxnRunner and a worker.Worker interface.
+// This is local to the package, allowing for better testing of the underlying
+// trackedDB worker.
+type TrackedDB interface {
+	coredatabase.TxnRunner
+	worker.Worker
+}
+
+// TrackedDBWorkerOption is a function that configures a TrackedDBWorker.
+type TrackedDBWorkerOption func(*trackedDBWorker)
+
+// WithClock sets the clock used by the worker.
+func WithClock(clock clock.Clock) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.clock = clock
+	}
+}
+
+// WithLogger sets the logger used by the worker.
+func WithLogger(logger logger.Logger) TrackedDBWorkerOption {
+	return func(w *trackedDBWorker) {
+		w.logger = logger
+	}
+}
+
+type trackedDBWorker struct {
+	tomb tomb.Tomb
+
+	dbApp     DBApp
+	namespace string
+
+	db    *sqlair.DB
+	rawDB *sql.DB
+
+	clock  clock.Clock
+	logger logger.Logger
+}
+
+// NewTrackedDBWorker creates a new TrackedDBWorker
+func NewTrackedDBWorker(
+	ctx context.Context, dbApp DBApp, namespace string, opts ...TrackedDBWorkerOption,
+) (TrackedDB, error) {
+	return newTrackedDBWorker(ctx, nil, dbApp, namespace, opts...)
+}
+
+func newTrackedDBWorker(
+	ctx context.Context, internalStates chan string,
+	dbApp DBApp, namespace string,
+	opts ...TrackedDBWorkerOption,
+) (TrackedDB, error) {
+	w := &trackedDBWorker{
+		dbApp:     dbApp,
+		namespace: namespace,
+		clock:     clock.WallClock,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	db, err := w.dbApp.Open(ctx, w.namespace)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	db.SetMaxIdleConns(1)
+	db.SetMaxOpenConns(1)
+
+	// Ensure that foreign keys are enabled, as we rely on them for
+	// referential integrity.
+	if err := pragma.SetPragma(ctx, db, pragma.ForeignKeysPragma, true); err != nil {
+		return nil, errors.Annotate(err, "setting foreign keys pragma")
+	}
+
+	w.rawDB = db
+	w.db = sqlair.NewDB(db)
```

Comment:
Any need for the SQLair transaction to be an option here?

Reply:
So we don't panic if someone invokes it on the txn runner. For such a cheap option, we gain a lot.

In theory, we **could** run sqlair commands in the repl OR, more likely, which I think would be amazing...

```sh
dbrepl (default)> .compile SELECT &something.* FROM table;
...
```

You'd have to work out how pass the "something" in to the query, but it should be doable.